Sources/Caching/DeviceCache.swift

```diff
@@ -57,14 +57,21 @@ class DeviceCache {
         self.userDefaults = .init(userDefaults: userDefaults)
         self.appUserIDHasBeenSet.value = userDefaults.string(forKey: .appUserDefaults) != nil
 
-        self.userDefaultsObserver = self.notificationCenter.addObserver(
-            forName: UserDefaults.didChangeNotification,
-            object: userDefaults,
-            queue: nil, // Run synchronously on the posting thread
-            using: { [weak self] notification in
-                self?.handleUserDefaultsChanged(notification: notification)
-            }
-        )
+        // Observe `UserDefaults` changes through `handleUserDefaultsChanged`
+        // to ensure that users don't remove the data from the SDK, which would
+        // leave it in an undetermined state. See https://rev.cat/userdefaults-crash
+        // If the user is not using a custom `UserDefaults`, we don't need to
+        // because they have no access to it.
```

Comment:
the good news is that folks who _are_ setting their custom user defaults for revenuecat usually know not to delete its contents, so the code will likely very rarely run